h2o-core/src/main/java/water/rapids/ast/prims/models/AstFairnessMetrics.java

```diff
@@ -0,0 +1,470 @@
+import water.util.ArrayUtils;
+import water.util.TwoDimTable;
+
+import java.lang.reflect.Field;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicIntegerArray;
+import java.util.stream.IntStream;
+
+public class AstFairnessMetrics extends AstPrimitive {
+    static public class FairnessMetrics {
+        double TP;
+        double FP;
+        double TN;
+        double FN;
+
+        double total;
+        double relativeSize;
+        double accuracy;
+        double precision;
+        double f1;
+        double sensitivity;
+        double specificity;
+        double falsePositiveRate;
+        double falseNegativeRate;
+        double auc;
+        double aucpr;
+        double gini;
+        double selected;
+        double selectedRatio;
+        double logloss;
+
+        public FairnessMetrics(double TP, double TN, double FP, double FN, double LLS, AUC2.AUCBuilder aucBuilder, double nrows) {
+            this.TP = TP;
+            this.TN = TN;
+            this.FP = FP;
+            this.FN = FN;
+            total = TP + FP + TN + FN;
+            logloss = -LLS/total;
+            relativeSize = total / nrows;
+            accuracy = (TP + TN)/total;
+            precision = TP / (FP + TP);
+            f1 = (2 * TP) / (2 * TP + FP + FN);
+            sensitivity = TP / (TP + FN);
+            specificity = TN / (TN + FP);
+            falsePositiveRate = FP / (FP + TN);
+            falseNegativeRate = FN / (FN + TP);
+            if (aucBuilder != null) {
+                AUC2 auc2 = new AUC2(aucBuilder);
+                auc = auc2._auc;
+                aucpr = auc2._pr_auc;
+                gini = auc2._gini;
+            } else {
+                auc = Double.NaN;
+                aucpr = Double.NaN;
+                gini = Double.NaN;
+            }
+            selected = TP + FP;
+            selectedRatio = (TP + FP) / total;
+        }
+    }
+
+    public static class FairnessMRTask extends MRTask {
+        int[] protectedColsIdx;
+        int[] cardinalities;
+        int responseIdx;
+        int predictionIdx;
+        final int TP = 0;
+        final int TN = 1;
+        final int FP = 2;
+        final int FN = 3;
+        final int LLS = 4; // Log Loss Sum
+        final int essentialMetrics = 5;
+        final int maxIndex;
+        final int favourableClass;
+
+        int[] _results;
+        AUC2.AUCBuilder[] _aucs;
+        public FairnessMRTask(int[] protectedColsIdx, int[] cardinalities, int responseIdx, int predictionIdx, int favourableClass) {
+            super();
+            this.protectedColsIdx = protectedColsIdx;
+            this.cardinalities = cardinalities;
+            this.responseIdx = responseIdx;
+            this.predictionIdx = predictionIdx;
+            this.favourableClass = favourableClass;
+            this.maxIndex = Arrays.stream(cardinalities).reduce((a,b) -> a*b ).getAsInt();
```

Comment:
>Are cardinalities equivalent to domain size for each protected columns?
If so, I don't expect them to ever be very large, but better safe than sorry.

Yes, but better safe than sorry. This was already checked before invocation of this constructor but it's better to check it here as well just in case somebody would want to reuse it.

The check before the invocation has limit of 1M (magic constant) and even with this number it would probably take too long to be practical... And interpreting 1M groups would also be unfeasible. But in theory I could work with `Integer.MAX_VALUE` groups which is what I'll check for in the constructor but I'd keep the check in the invocation for the 1M or some smaller value to make sure we don't accidentally get stuck (e.g., passing continuous column encoded as factor instead of categorical column etc). But I don't have a strong opinion about that.